// mcp-server-gsc/dist/webmaster.js
import { google } from 'googleapis';
export class WebmasterService {
    auth;
    constructor(credentials) {
        this.auth = new google.auth.GoogleAuth({
            keyFile: credentials,
            scopes: ['https://www.googleapis.com/auth/webmasters.readonly'],
        });
    }
    async getWebmasters() {
        const authClient = await this.auth.getClient();
        return google.webmasters({
            version: 'v3',
            auth: authClient,
        });
    }
    async getSearchConsole() {
        const authClient = await this.auth.getClient();
        return google.searchconsole({
            version: 'v1',
            auth: authClient,
        });
    }
    async searchAnalytics(siteUrl, requestBody) {
        const webmasters = await this.getWebmasters();
        try {
            const response = await webmasters.searchanalytics.query({
                siteUrl,
                requestBody,
            });
            return response;
        }
        catch (err) {
            if (err instanceof Error && err.message.includes('sufficient permission for site')) {
                let url = new URL(siteUrl);
                if (url.protocol === 'http:' || url.protocol === 'https:') {
                    url = new URL('sc-domain:' + url.hostname);
                }
                else {
                    url = new URL('https://' + url.toString());
                }
                const response = await webmasters.searchanalytics.query({
                    siteUrl: siteUrl.toString(),
                    requestBody,
                });
                return response;
            }
            throw err;
        }
    }
    async listSites() {
        const webmasters = await this.getWebmasters();
        const response = await webmasters.sites.list();
        return response;
    }
    async indexInspect(requestBody) {
        const searchConsole = await this.getSearchConsole();
        const response = await searchConsole.urlInspection.index.inspect({
            requestBody,
        });
        return response;
    }
}
